Simplenote/src/main/java/com/automattic/simplenote/TagDialogFragment.java

```diff
@@ -81,80 +103,60 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
 
-                int index = mTag.hasIndex() ? mTag.getIndex() : mBucketTag.count();
-                boolean isRenamingToLexicalTag = TagUtils.hashTag(tagNew).equals(TagUtils.hashTag(mTagOld));
-                boolean hasCanonicalTag = TagUtils.hasCanonicalOfLexical(mBucketTag, tagNew);
+        mClickListenerPositiveConflict = v -> viewModel.renameTag();
+    }
 
-                if (hasCanonicalTag && !isRenamingToLexicalTag) {
-                    String tagCanonical = TagUtils.getCanonicalFromLexical(mBucketTag, tagNew);
-                    showDialogErrorConflict(tagCanonical, mTagOld);
-                    return;
-                }
 
-                tryToRenameTag(tagNew, index);
+    private void setObservers() {
+        viewModel.getUiState().observe(this, uiState -> {
+            // Validate if the current state has an error
+            if (uiState.getErrorMsg() != null) {
+                mEditTextLayout.setError(getString(uiState.getErrorMsg()));
+                mButtonPositive.setEnabled(false);
+            } else {
+                // If there is not an error, enable save button
+                mEditTextLayout.setError(null);
+                mButtonPositive.setEnabled(true);
             }
-        };
+        });
 
-        mClickListenerNeutralConflict = new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                showDialogRenameTag();
-            }
-        };
-
-        mClickListenerPositiveConflict = new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                String tagNew = getTagNew();
-                int index = mTag.hasIndex() ? mTag.getIndex() : mBucketTag.count();
-                tryToRenameTag(tagNew, index);
+        viewModel.getEvent().observe(this, event -> {
+            if (event instanceof CloseEvent || event instanceof FinishEvent) {
+                dismiss();
+            } else if (event instanceof ShowErrorEvent) {
+                Context context = requireContext();
+                DialogUtils.showDialogWithEmail(
+                        context,
+                        context.getString(R.string.rename_tag_message)
+                );
+            } else if (event instanceof ConflictEvent) {
+                ConflictEvent conflictEvent = (ConflictEvent) event;
+                showDialogErrorConflict(conflictEvent.getCanonicalTagName(), conflictEvent.getOldTagName());
             }
-        };
-
-        return mDialogEditTag;
+        });
     }
 
     @Override
     public void onShow(DialogInterface dialog) {
         mButtonNegative = mDialogEditTag.getButton(DialogInterface.BUTTON_NEGATIVE);
         mButtonNeutral = mDialogEditTag.getButton(DialogInterface.BUTTON_NEUTRAL);
         mButtonPositive = mDialogEditTag.getButton(DialogInterface.BUTTON_POSITIVE);
+
+        // Set observers when views are available
+        setObservers();
```

Comment:
Question (❓): Why is `setObservers()` being trigger during `onShow(...)` instead of `onCreate(...)`, and more specifically before the call to `viewModel.start(mTag)`?

Reply:
In `onCreate(...)` some views that are specific to the dialog (e,g, positive, negative buttons) are not available (they return `null` if we want to get them). The observers use these views specifically, therefore it would throw a null pointer exception. What I did change was the call to `viewModelo.start(...)`, I put it after I call the observers.

Reply:
Brilliant, thanks for doing so @dandominguez04 !